fix: outline appearing over child

iamdipanshusingh opened the pull request:
<!--
Thanks for filing a pull request!
Reviewers are typically assigned within a week of filing a request.
To learn more about code review, see our documentation on Tree Hygiene: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
-->

This PR exposes `borderOnForeground` from the `Material` class to `ButtonStyle` and `OutlinedButton` class, letting the developer decide whether to show the border on foreground or background.

Screenshot before this PR:
<img width="212" alt="before" src="https://github.com/user-attachments/assets/5b23cdf8-bbd7-4bee-9b15-ee64aa941043" />

Screenshot after this PR:
<img width="200" alt="after" src="https://github.com/user-attachments/assets/1d3f66c8-c2f2-4942-a92e-d890d3c65a22" />


fixes #167431 

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

iamdipanshusingh committed:
fix: outline appearing over child

iamdipanshusingh committed:
chore: borderOnForeground defaults to null

iamdipanshusingh committed:
test: added relevant tests

iamdipanshusingh commented:
here's the code sample to make the border appear behind the child

```dart
Column(
  mainAxisSize: MainAxisSize.min,
  children: [
    IconButton.outlined(
      iconSize: 40,
      isSelected: false,
      onPressed: () {},
      borderOnForeground: false,
      icon: Badge(
        backgroundColor: Theme.of(context).colorScheme.primary,
        label: Text('Ad', style: TextStyle(fontSize: 18)),
        child: Icon(Icons.lightbulb_rounded),
      ),
    ),
    const SizedBox(height: 20),
    SizedBox(
      height: 30,
      child: OutlinedButton(
        onPressed: () {},
        style: ButtonStyle(
          borderOnForeground: false,
        ),
        child: Badge(
          backgroundColor: Theme.of(context).colorScheme.primary,
          label: Text('Ad', style: TextStyle(fontSize: 18)),
          child: Icon(Icons.lightbulb_rounded),
        ),
      ),
    ),
  ],
),
```

dkwingsmt commented:
Instead of making it an option, shall we treat this as a bug and make border-on-background the only behavior? The filled button's filling is always behind the child, and therefore it makes sense for the border of outlined buttons to always be as well.

iamdipanshusingh commented:
makes sense, I'll make the chsnges

iamdipanshusingh committed:
chore: border on foreground now defaults to false

iamdipanshusingh committed:
Merge branch 'master' into ds/fix/issue-167431

iamdipanshusingh commented:
@dkwingsmt I've made the changes, `borderOnForeground` now defaults to false in `ButtonStyleButton`.

iamdipanshusingh commented:
also @dkwingsmt what am I doing wrong here? why are so much tests are failing?

dkwingsmt commented:
```
09:04 +935 ~359: test/material/outlined_button_test.dart: OutlinedButton, OutlinedButton.icon defaults
```

This test seems to also need a fix.
`packages/flutter/test/material/outlined_button_test.dart` L44


Also can you add another test to verify the appearance of a outlined button with a badge as a regression test? Either a `paint` verifier or a golden test would be ok.

iamdipanshusingh committed:
Merge branch 'master' into ds/fix/issue-167431

iamdipanshusingh committed:
fix: borderOnForeground test

iamdipanshusingh committed:
added golden test for outlined button and icon button

iamdipanshusingh committed:
fix: button default tests

iamdipanshusingh committed:
chore: renamed test

iamdipanshusingh commented:
@dkwingsmt I've added the golden tests as requested, attaching the screenshots below.
I also noticed that other buttons' default tests were failing, and I've fixed that as well.

![material icon_button badge outline](https://github.com/user-attachments/assets/999d1883-e071-4db6-81cf-71c4ff8eeac3)
![material outlined_button badge outline](https://github.com/user-attachments/assets/ba2c778f-9e8e-4a4c-8c24-c4862f25003d)

---

Let me know if any further changes are required

iamdipanshusingh committed:
chore: added tags 'reduced-test-set' for golden tests

iamdipanshusingh committed:
Merge branch 'master' into ds/fix/issue-167431

dkwingsmt commented on packages/flutter/test/material/icon_button_test.dart:
```diff
@@ -3375,6 +3381,34 @@ void main() {
     expect(inkFeatures, paints..rect(color: focusColor));
   }, skip: !isBrowser); // [intended] tests web-specific behavior.
+
+  testWidgets('IconButton golden test', (WidgetTester tester) async {
```
```suggestion
  testWidgets('IconButton with a badge', (WidgetTester tester) async {
```

iamdipanshusingh replied:
@dkwingsmt is it a typo here? `Outlined button's outline` shouldn't this be `IconButton's outline`?

iamdipanshusingh replied:
@dkwingsmt BTW renamed the test, and added the link to the issue

dkwingsmt commented:
Also the analyzer is failing due to format errors, which can be fixed by running `dart fix .` under `packages/flutter`.

dkwingsmt reviewed (approved):
LGTM otherwise! Great change, thank you!

iamdipanshusingh commented:
> Also the analyzer is failing due to format errors, which can be fixed by running `dart fix .` under `packages/flutter`.

it says nothing to fix(in `packages/flutter`) 

<img width="644" alt="Screenshot 2025-04-28 at 12 35 40 PM" src="https://github.com/user-attachments/assets/b3154179-2cb7-42ef-9544-e7d75cf2ab8f" />

---

but proposed 313 fixes in the root folder, which I think is out of scope of this PR 😅 

<img width="632" alt="Screenshot 2025-04-28 at 12 39 33 PM" src="https://github.com/user-attachments/assets/f0cb0824-bb2a-421f-8556-cd806bfbbace" />

---

@dkwingsmt  can we merge this PR without it?

iamdipanshusingh committed:
chore: rename test, added issue link

dkwingsmt commented:
We have to make the formatter happy though... not sure why it doesn't work, but can you manually fix the format as indicated?
```

                                                                                                    
Found 1 Dart file which was formatted incorrectly.
To fix, run `dart format packages/flutter/test/material/icon_button_test.dart` or:

git apply <<DONE
diff --git a/packages/flutter/test/material/icon_button_test.dart b/packages/flutter/test/material/icon_button_test.dart
index 4f9abbe975..0000000000 100644
--- a/packages/flutter/test/material/icon_button_test.dart
+++ b/packages/flutter/test/material/icon_button_test.dart
@@ -7,7 +7,6 @@
 @Tags(<String>['reduced-test-set'])
 library;
 
-
 import 'dart:ui';
 
 import 'package:flutter/foundation.dart';
@@ -3405,10 +3404,7 @@ void main() {
     );
 
     await tester.pumpAndSettle();
-    await expectLater(
-      find.byType(IconButton),
-      matchesGoldenFile('icon_button.badge.outline.png'),
-    );
+    await expectLater(find.byType(IconButton), matchesGoldenFile('icon_button.badge.outline.png'));
   });
 }
 
@@ -3480,3 +3476,4 @@ TextStyle? _iconStyle(WidgetTester tester, IconData icon) {
   );
   return iconRichText.text.style;
 }
DONE
```

iamdipanshusingh committed:
chore: format code

iamdipanshusingh committed:
Merge branch 'master' into ds/fix/issue-167431

iamdipanshusingh commented:
@dkwingsmt can you check now? I've formatted the file.
But the tree status is failing. I also updated the branch (merge commit), but it's still not fixed. Let me know how I can fix that.

justinmc commented on packages/flutter/test/material/outlined_button_test.dart:
```diff
@@ -2954,4 +2961,33 @@ void main() {
+      ),
+    );
+
+    await tester.pumpAndSettle();
```
Same question about this pumpAndSettle.

justinmc reviewed (approved):
LGTM with a nit about possible overuse of pumpAndSettle :+1: 

Thank you for jumping in and fixing this!

iamdipanshusingh commented on packages/flutter/test/material/icon_button_test.dart:
```diff
@@ -3375,6 +3380,32 @@ void main() {
+      ),
+    );
+
+    await tester.pumpAndSettle();
```
yeah, seems like it 😅 removing it

iamdipanshusingh committed:
chore: remove unnecessary use of pumpAndSettle

iamdipanshusingh commented:
@justinmc removed unnecessary `pumpAndSettle`
Let me know if anything else is required 😄

dkwingsmt commented:
I've checked Google testing failures and all of them are caused by swapping the order between the content and the outline of buttons. They're expected and in my opinion negligible and acceptable.
